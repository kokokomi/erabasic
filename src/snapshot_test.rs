use std::fs;
use pest::Parser;
use crate::{EraBasicParser, Rule};
use insta::assert_debug_snapshot;

#[test]
fn functions() {
    let unparsed_file = fs::read_to_string("example/functions.erb").expect("cannot read file");
    let file = EraBasicParser::parse(Rule::file, &unparsed_file)
        .expect("unsuccessful parse");
    assert_debug_snapshot!(file);
}

#[test]
fn if_statement() {
    let unparsed_file = fs::read_to_string("example/if_statement.erb").expect("cannot read file");
    let file = EraBasicParser::parse(Rule::file, &unparsed_file)
        .expect("unsuccessful parse");
        
    assert_debug_snapshot!(file);
}

#[test]
fn sif_statement() {
    let unparsed_file = fs::read_to_string("example/sif_statement.erb").expect("cannot read file");
    let file = EraBasicParser::parse(Rule::file, &unparsed_file)
        .expect("unsuccessful parse");

    assert_debug_snapshot!(file);
}

#[test]
fn repeat_statement() {
    let unparsed_file = fs::read_to_string("example/repeat_statement.erb").expect("cannot read file");
    let file = EraBasicParser::parse(Rule::file, &unparsed_file)
        .expect("unsuccessful parse");

    assert_debug_snapshot!(file);
}

#[test]
fn while_statement() {
    let unparsed_file = fs::read_to_string("example/while_statement.erb").expect("cannot read file");
    let file = EraBasicParser::parse(Rule::file, &unparsed_file)
        .expect("unsuccessful parse");

    assert_debug_snapshot!(file);
}

#[test]
fn do_loop_statement() {
    let unparsed_file = fs::read_to_string("example/do_loop_statement.erb").expect("cannot read file");
    let file = EraBasicParser::parse(Rule::file, &unparsed_file)
        .expect("unsuccessful parse");

    assert_debug_snapshot!(file);
}