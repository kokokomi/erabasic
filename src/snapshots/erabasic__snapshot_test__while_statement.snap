---
source: src/snapshot_test.rs
expression: file
---
[
    Pair {
        rule: function_declare,
        span: Span {
            str: "@while_statement\n  while 1\n    if 1 == 1\n      continue\n    else\n      break\n    endif\n  wend\n",
            start: 0,
            end: 94,
        },
        inner: [
            Pair {
                rule: function_identifier,
                span: Span {
                    str: "while_statement",
                    start: 1,
                    end: 16,
                },
                inner: [
                    Pair {
                        rule: identifier,
                        span: Span {
                            str: "while_statement",
                            start: 1,
                            end: 16,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: function_body,
                span: Span {
                    str: "while 1\n    if 1 == 1\n      continue\n    else\n      break\n    endif\n  wend\n",
                    start: 19,
                    end: 94,
                },
                inner: [
                    Pair {
                        rule: statement,
                        span: Span {
                            str: "while 1\n    if 1 == 1\n      continue\n    else\n      break\n    endif\n  wend",
                            start: 19,
                            end: 93,
                        },
                        inner: [
                            Pair {
                                rule: while_statement,
                                span: Span {
                                    str: "while 1\n    if 1 == 1\n      continue\n    else\n      break\n    endif\n  wend",
                                    start: 19,
                                    end: 93,
                                },
                                inner: [
                                    Pair {
                                        rule: while_keyword,
                                        span: Span {
                                            str: "while",
                                            start: 19,
                                            end: 24,
                                        },
                                        inner: [],
                                    },
                                    Pair {
                                        rule: expression,
                                        span: Span {
                                            str: "1",
                                            start: 25,
                                            end: 26,
                                        },
                                        inner: [
                                            Pair {
                                                rule: primary,
                                                span: Span {
                                                    str: "1",
                                                    start: 25,
                                                    end: 26,
                                                },
                                                inner: [
                                                    Pair {
                                                        rule: integer_literal,
                                                        span: Span {
                                                            str: "1",
                                                            start: 25,
                                                            end: 26,
                                                        },
                                                        inner: [],
                                                    },
                                                ],
                                            },
                                        ],
                                    },
                                    Pair {
                                        rule: block,
                                        span: Span {
                                            str: "if 1 == 1\n      continue\n    else\n      break\n    endif\n",
                                            start: 31,
                                            end: 87,
                                        },
                                        inner: [
                                            Pair {
                                                rule: statement,
                                                span: Span {
                                                    str: "if 1 == 1\n      continue\n    else\n      break\n    endif",
                                                    start: 31,
                                                    end: 86,
                                                },
                                                inner: [
                                                    Pair {
                                                        rule: if_statement,
                                                        span: Span {
                                                            str: "if 1 == 1\n      continue\n    else\n      break\n    endif",
                                                            start: 31,
                                                            end: 86,
                                                        },
                                                        inner: [
                                                            Pair {
                                                                rule: if_keyword,
                                                                span: Span {
                                                                    str: "if",
                                                                    start: 31,
                                                                    end: 33,
                                                                },
                                                                inner: [],
                                                            },
                                                            Pair {
                                                                rule: expression,
                                                                span: Span {
                                                                    str: "1 == 1",
                                                                    start: 34,
                                                                    end: 40,
                                                                },
                                                                inner: [
                                                                    Pair {
                                                                        rule: binary_expression,
                                                                        span: Span {
                                                                            str: "1 == 1",
                                                                            start: 34,
                                                                            end: 40,
                                                                        },
                                                                        inner: [
                                                                            Pair {
                                                                                rule: primary,
                                                                                span: Span {
                                                                                    str: "1",
                                                                                    start: 34,
                                                                                    end: 35,
                                                                                },
                                                                                inner: [
                                                                                    Pair {
                                                                                        rule: integer_literal,
                                                                                        span: Span {
                                                                                            str: "1",
                                                                                            start: 34,
                                                                                            end: 35,
                                                                                        },
                                                                                        inner: [],
                                                                                    },
                                                                                ],
                                                                            },
                                                                            Pair {
                                                                                rule: operator,
                                                                                span: Span {
                                                                                    str: "==",
                                                                                    start: 36,
                                                                                    end: 38,
                                                                                },
                                                                                inner: [
                                                                                    Pair {
                                                                                        rule: conditional_operator,
                                                                                        span: Span {
                                                                                            str: "==",
                                                                                            start: 36,
                                                                                            end: 38,
                                                                                        },
                                                                                        inner: [
                                                                                            Pair {
                                                                                                rule: equals_op,
                                                                                                span: Span {
                                                                                                    str: "==",
                                                                                                    start: 36,
                                                                                                    end: 38,
                                                                                                },
                                                                                                inner: [],
                                                                                            },
                                                                                        ],
                                                                                    },
                                                                                ],
                                                                            },
                                                                            Pair {
                                                                                rule: expression,
                                                                                span: Span {
                                                                                    str: "1",
                                                                                    start: 39,
                                                                                    end: 40,
                                                                                },
                                                                                inner: [
                                                                                    Pair {
                                                                                        rule: primary,
                                                                                        span: Span {
                                                                                            str: "1",
                                                                                            start: 39,
                                                                                            end: 40,
                                                                                        },
                                                                                        inner: [
                                                                                            Pair {
                                                                                                rule: integer_literal,
                                                                                                span: Span {
                                                                                                    str: "1",
                                                                                                    start: 39,
                                                                                                    end: 40,
                                                                                                },
                                                                                                inner: [],
                                                                                            },
                                                                                        ],
                                                                                    },
                                                                                ],
                                                                            },
                                                                        ],
                                                                    },
                                                                ],
                                                            },
                                                            Pair {
                                                                rule: block,
                                                                span: Span {
                                                                    str: "continue\n",
                                                                    start: 47,
                                                                    end: 56,
                                                                },
                                                                inner: [
                                                                    Pair {
                                                                        rule: statement,
                                                                        span: Span {
                                                                            str: "continue",
                                                                            start: 47,
                                                                            end: 55,
                                                                        },
                                                                        inner: [
                                                                            Pair {
                                                                                rule: continue_keyword,
                                                                                span: Span {
                                                                                    str: "continue",
                                                                                    start: 47,
                                                                                    end: 55,
                                                                                },
                                                                                inner: [],
                                                                            },
                                                                        ],
                                                                    },
                                                                ],
                                                            },
                                                            Pair {
                                                                rule: else_statement,
                                                                span: Span {
                                                                    str: "else\n      break\n",
                                                                    start: 60,
                                                                    end: 77,
                                                                },
                                                                inner: [
                                                                    Pair {
                                                                        rule: else_keyword,
                                                                        span: Span {
                                                                            str: "else",
                                                                            start: 60,
                                                                            end: 64,
                                                                        },
                                                                        inner: [],
                                                                    },
                                                                    Pair {
                                                                        rule: block,
                                                                        span: Span {
                                                                            str: "break\n",
                                                                            start: 71,
                                                                            end: 77,
                                                                        },
                                                                        inner: [
                                                                            Pair {
                                                                                rule: statement,
                                                                                span: Span {
                                                                                    str: "break",
                                                                                    start: 71,
                                                                                    end: 76,
                                                                                },
                                                                                inner: [
                                                                                    Pair {
                                                                                        rule: break_keyword,
                                                                                        span: Span {
                                                                                            str: "break",
                                                                                            start: 71,
                                                                                            end: 76,
                                                                                        },
                                                                                        inner: [],
                                                                                    },
                                                                                ],
                                                                            },
                                                                        ],
                                                                    },
                                                                ],
                                                            },
                                                            Pair {
                                                                rule: endif_keyword,
                                                                span: Span {
                                                                    str: "endif",
                                                                    start: 81,
                                                                    end: 86,
                                                                },
                                                                inner: [],
                                                            },
                                                        ],
                                                    },
                                                ],
                                            },
                                        ],
                                    },
                                    Pair {
                                        rule: wend_keyword,
                                        span: Span {
                                            str: "wend",
                                            start: 89,
                                            end: 93,
                                        },
                                        inner: [],
                                    },
                                ],
                            },
                        ],
                    },
                ],
            },
        ],
    },
    Pair {
        rule: EOI,
        span: Span {
            str: "",
            start: 94,
            end: 94,
        },
        inner: [],
    },
]
